File: CircularLinkedListPython/CircularLinkedList.py
class CircularLinkedList:
    class Node:
        def __init__(self, data = None):
            self.data = data
            self.next = None

    def __init__(self, size):
        self.__size = size
        self.__count = -1
        self.__head = self.Node()

    def __is_list_empty(self):
        return self.__count == -1

    def __is_list_full(self):
        return self.__count == self.__size - 1
    
    def insertion_at_the_beginning(self, newData):
        if self.__is_list_full():
            print("Your List is Full\n")
            return

        newNode = self.Node(newData)

        newNode.next = self.__head

        node = self.__head

        for _ in range(self.__count):
            node = node.next

        node.next = newNode

        self.__head = newNode

        self.__count += 1        


    def insertion_at_the_end(self, newData):
        if self.__is_list_full():
            print("Your List is Full\n")
            return
        elif self.__is_list_empty():
            self.insertion_at_the_beginning(newData)
            return
        
        newNode = self.Node(newData)

        node = self.__head

        for _ in range(self.__count):
            node = node.next

        newNode.next = self.__head

        node.next = newNode

        self.__count += 1    

    def insertion_after_node(self, newData, nodePos):
        if self.__is_list_full():
            print("Your List is Full\n")
            return
        elif nodePos < 1:
            self.insertion_at_the_beginning(newData)
            return
        elif nodePos - 1 >= self.__count:
            self.insertion_at_the_end(newData)
            return
        else:
            newNode = self.Node(newData)

            node = self.__head

            for _ in range(nodePos - 1):
                node = node.next

            newNode.next = node.next
            node.next = newNode

            self.__count += 1

    def delete_at_the_beginning(self):
        if self.__is_list_empty(): 
            print("Your List is Empty!\n")
            return

        self.__head = self.__head.next

        self.__count -= 1     
    
    def delete_at_the_end(self):
        if self.__is_list_empty(): 
            print("Your List is Empty!\n")
            return
        
        node = self.__head

        for _ in range(self.__count - 1):
            node = node.next

        node.next = self.__head

        self.__count -= 1

    def delete_node(self, nodePos):
        if self.__is_list_empty(): 
            print("Your List is Empty!\n")
            return
        elif nodePos == 0:
            self.delete_at_the_beginning()
            return
        elif nodePos - 1 >= self.__count and nodePos <= self.__size:
            self.delete_at_the_end()
            return
        elif nodePos > 0 and nodePos - 1 < self.__count:
            node = self.__head
            prev = None

            for _ in range(nodePos):
                prev = node
                node = node.next

            if prev != None:
                prev.next = node.next
            else:
                self.__head = None

            self.__count -= 1
        else:
            print('Invalid position to delete!\n')
            return

    def search_node(self, value):
        if self.__is_list_empty(): 
            print("Your List is Empty!\n")
            return
        
        node = self.__head

        for _ in range(self.__count):
            if node.data == value: return True
            else: node = node.next

        return True
        


    def display_list(self):
        if self.__is_list_empty(): 
            print("Your List is Empty!\n")
            return
    
        node = self.__head
        print(f"Your List is: {node.data}", end='')

        node = node.next
        for _ in range(self.__count):
            print(f'->{node.data}', end='')
            node = node.next
        print('\n')

File: CircularLinkedListCpp/CircularLinkedListCpp/CircularLinkedList.cpp
#include <iostream>

#include "CircularLinkedList.hpp"

#define SIZE 5

bool CircularLinkedList::__isListEmpty() {
	return (__head == nullptr);
}

bool CircularLinkedList::__isListFull() {
	return (__count == SIZE - 1);
}

CircularLinkedList::CircularLinkedList() { __head = new Node; }

CircularLinkedList::~CircularLinkedList() {
	Node* node = __head;
	Node* next = nullptr;
	for (int i = 0; i < __count; i++) {
		next = node->next;
		delete node;
		node = next;
	}

	__head = nullptr;
}

void CircularLinkedList::insertionAtTheBeggining(int newData) {
	if (__isListFull()) {
		std::cout << "Your List is Full!" << std::endl;
		return;
	}

	Node* newNode = new Node;
	newNode->data = newData;
	newNode->next = __head;

	Node* node = __head;

	for (int i = 0; i < __count; i++) {
		node = node->next;
	}

	node->next = newNode;

	__head = newNode;

	++__count;
}

void CircularLinkedList::insertionAtTheEnd(int newData) {
	if (__isListFull()) {
		std::cout << "Your List is Full!" << std::endl;
		return;
	}
	else if (__isListEmpty()) {
		insertionAtTheBeggining(newData);
		return;
	}
	else {
		Node* newNode = new Node;
		newNode->data = newData;

		Node* node = __head;

		for (int i = 0; i < __count; i++) {
			node = node->next;
		}
		newNode->next = __head;

		node->next = newNode;

		++__count;
	}
}

void CircularLinkedList::insertionAfterNode(int newData, int nodePos) {
	if (__isListFull()) {
		std::cout << "Your List is Full!" << std::endl;
		return;
	}
	else if (nodePos < 1) {
		insertionAtTheBeggining(newData);
		return;
	}
	else if (nodePos - 1 >= __count) {
		insertionAtTheEnd(newData);
		return;
	}
	else {
		Node* newNode = new Node;
		newNode->data = newData;

		Node* node = __head;

		for (int i = 0; i < nodePos - 1; i++) {
			node = node->next;
		}

		newNode->next = node->next;
		node->next = newNode;

		++__count;
	}

}

void CircularLinkedList::deleteAtTheBeggining() {
	if (__isListEmpty()) {
		std::cout << "Your List is Empty!" << std::endl;
		return;
	}

	Node* node = __head;
	__head = __head->next;

	delete node;

	--__count;
}

void CircularLinkedList::deleteAtTheEnd() {
	if (__isListEmpty()) {
		std::cout << "Your List is Empty!" << std::endl;
		return;
	}
	else if (__count == 0) {
		Node* node = __head;
		delete node;
		__head = nullptr;
		--__count;
	}
	else {
		Node* node = __head;

		for (int i = 0; i < __count - 1; i++) {
			node = node->next;
		}

		delete node->next;
		node->next = __head;

		--__count;
	}

}

void CircularLinkedList::deleteNode(int nodePos) {
	if (__isListEmpty()) {
		std::cout << "Your List is Empty!" << std::endl;
		return;
	} else if (nodePos == 0) {
		deleteAtTheBeggining();
		return;
	}
	else if (nodePos - 1 >= __count && nodePos <= SIZE) {
		deleteAtTheEnd();
		return;
	}
	else if (nodePos > 0 && nodePos < SIZE) {
		Node* node = __head;
		Node* prev = nullptr;

		for (int i = 0; i < nodePos - 1; i++) {
			prev = node;
			node = node->next;
		}

		if (prev != nullptr) {
			prev->next = node->next;
			delete node;
		}
		else {
			delete __head;
			__head = nullptr;
		}

		--__count;

	}
	else {
		std::cout << "\nInvalid Node Position to delete!\n";
		return;
	}
}

bool CircularLinkedList::searchNode(int data) {
	if (__isListEmpty()) {
		std::cout << "Your List is Empty!" << std::endl;
		return false;
	}

	Node* node = __head;

	for (int i = 0; i <= __count; i++) {
		if (node->data == data) {
			return true;
		}
		node = node->next;
	}

	return false;
}

void CircularLinkedList::displayList() {
	if (__isListEmpty()) {
		std::cout << "Your List is Empty!" << std::endl;
		return;
	}

	Node* node = __head;
	std::cout << "Your List is: " << __head->data;

	node = node->next;

	for (int i = 0; i < __count; i++) {
		std::cout << "->" << node->data;
		node = node->next;
	}

	std::cout << std::endl;
}